### PR TITLE
Feature/storage/with version encoding bug

### LINKED ISSUE #13222: [BUG] BlobBaseClient.WithVersion()  will encode blob name
**Describe the bug**
In the just released Azure.Storage.Blobs 12.5.0-preview.5, the BlobBaseClient.WithVersion() not work well.

If the Blob Client has blob name include any special char like “/”, the char will be encoded after run BlobBaseClient.WithVersion().

I believe BlobBaseClient.WithSnapshot() should have same problem.
This issue also happen on class inherit from BlobBaseClient, like BlockBlobClient.WithVersion().

**Expected behavior**
the blob name should not be encoded after run BlobBaseClient.WithVersion() 

**Actual behavior (include Exception or Stack Trace)**
What is the actual behavior?
image

**To Reproduce**
1. Create a blobbaseclient with base blob Uri, the blob name contains "/", like "dir1/blob1"
2. run blobbaseclient.WithVersion(versionId) , the blob name encoded to "dir1%2Fblob1"

**Environment:**
 - Name and version of the Library package used: [e.g. Azure.Storage.Blobs 12.5.0-preview.5] 
 - Hosting platform or OS and .NET runtime version (`dotnet --info` output for .NET Core projects): [e.g. Azure AppService or Windows 10 .NET Framework 4.8]
```
.NET Core SDK (reflecting any global.json):
 Version:   2.2.108
 Commit:    33ed5b90ce

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.18363
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.2.108\

Host (useful for support):
  Version: 2.2.6
  Commit:  7dac9b1b51

.NET Core SDKs installed:
  2.2.108 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.2.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.2.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.2.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download
```
 - IDE and version : [e.g. Visual Studio 16.3]
VS 15.7.4

### PR DESCRIPTION
**Pull request by seanmcc-msft**
Resolves https://github.com/Azure/azure-sdk-for-net/issues/13222

### CODE REVIEW

File: sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs [hunk 1]
```diff
@@ -32,6 +32,8 @@ public partial class BlobClient : Azure.Storage.Blobs.Specialized.BlobBaseClient
         public virtual System.Threading.Tasks.Task<Azure.Response<Azure.Storage.Blobs.Models.BlobContentInfo>> UploadAsync(string path, Azure.Storage.Blobs.Models.BlobUploadOptions options, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
         public virtual System.Threading.Tasks.Task<Azure.Response<Azure.Storage.Blobs.Models.BlobContentInfo>> UploadAsync(string path, bool overwrite = false, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
         public virtual System.Threading.Tasks.Task<Azure.Response<Azure.Storage.Blobs.Models.BlobContentInfo>> UploadAsync(string path, System.Threading.CancellationToken cancellationToken) { throw null; }
+        public new Azure.Storage.Blobs.BlobClient WithSnapshot(string snapshot) { throw null; }
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs:35]: BlobClient.WithSnapshot() and .WithVersion() don't currently work without this.

Comment (blueww) [reply sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs:35]: Do you mean  BlobClient.WithSnapshot() and .WithVersion() don't work after the fix?
But I see you have test case below for BlobClient on the 2 APIs.

File: sdk/storage/Azure.Storage.Blobs/src/BlobUriBuilder.cs [hunk 2]
```diff
@@ -215,7 +216,7 @@ public BlobUriBuilder(Uri uri)
                 else
                 {
                     BlobContainerName = path.Substring(startIndex, containerEndIndex - startIndex); // The container name is the part between the slashes
-                    BlobName = path.Substring(containerEndIndex + 1);   // The blob name is after the container slash
+                    BlobName = path.Substring(containerEndIndex + 1).UnescapePath();   // The blob name is after the container slash
```
Comment (blueww) [inline sdk/storage/Azure.Storage.Blobs/src/BlobUriBuilder.cs:219]: Will this function be called in other API?
If so, please cover other API's test with blob name contains special chars to make sure no regression.

File: sdk/storage/Azure.Storage.Blobs/tests/AppendBlobClientTests.cs [hunk 3]
```diff
@@ -99,59 +99,59 @@ public void Ctor_CPK_EncryptionScope()
...
-
-            builder = new BlobUriBuilder(blob.Uri);
-
-            Assert.AreEqual("foo", builder.Snapshot);
-
-            blob = InstrumentClient(blob.WithSnapshot(null));
+            // Arrange
+            string accountName = "accountname";
+            string containerName = GetNewContainerName();
+            string blobName = "my/blob/name";
```
Comment (blueww) [inline sdk/storage/Azure.Storage.Blobs/tests/AppendBlobClientTests.cs:105]: Please also add other special chars besides "/" in the blob name like 
%:?*:"<>|

File: sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs (hunk 1 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs:35]: They didn't work before the fix.
`BlobClient snapshotBlobClient = blobClient.WithSnapshot("...");`
`BlobClient versionBlobClient = blobClient.WithVersion("...");`
`Cannot implicitly convert type Azure.Storage.Blobs.Specialized.BaseBlobClient to Azure.Storage.Blobs.BlobClient...`

File: sdk/storage/Azure.Storage.Blobs/tests/AppendBlobClientTests.cs (hunk 3 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs/tests/AppendBlobClientTests.cs:105]: That case is covered by tests that were previously written.  https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/storage/Azure.Storage.Blobs/tests/ContainerClientTests.cs#L2256

File: sdk/storage/Azure.Storage.Blobs/src/BlobUriBuilder.cs (hunk 2 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs/src/BlobUriBuilder.cs:219]: This is covered by previously written tests - https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/storage/Azure.Storage.Blobs/tests/ContainerClientTests.cs#L2256

**Review by kasobol-msft — APPROVED** [review]

**Review by amnguye — APPROVED** [review]

File: sdk/storage/Azure.Storage.Blobs/tests/AppendBlobClientTests.cs (hunk 3 above)
Comment (blueww) [reply sdk/storage/Azure.Storage.Blobs/tests/AppendBlobClientTests.cs:126]: Would you like add WithVersion, withSanpshot test in this case?

File: sdk/storage/Azure.Storage.Blobs/src/BlobClient.cs [hunk 4]
```diff
@@ -182,6 +182,43 @@ public BlobClient(Uri blobUri, TokenCredential credential, BlobClientOptions opt
...
+        /// <param name="snapshot">The snapshot identifier.</param>
+        /// <returns>A new <see cref="BlobClient"/> instance.</returns>
+        /// <remarks>
+        /// Pass null or empty string to remove the snapshot returning a URL
+        /// to the base blob.
+        /// </remarks>
+        public new BlobClient WithSnapshot(string snapshot)
+        {
+            BlobUriBuilder blobUriBuilder = new BlobUriBuilder(Uri) { Snapshot = snapshot };
+            return new BlobClient(blobUriBuilder.ToUri(), Pipeline, Version, ClientDiagnostics, CustomerProvidedKey, ClientSideEncryption, EncryptionScope);
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Blobs/src/BlobClient.cs:201]: Why do we unescape the path in `BlobBaseClient` but not here in `BlobClient`?  If we should be doing it everywhere, is that work fixing automatically as part of `BlobUriBuilder`'s .ctor?

**Review by tg-msft — APPROVED** [review]

File: sdk/storage/Azure.Storage.Blobs/src/BlobClient.cs (hunk 4 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs/src/BlobClient.cs:201]: That was an oversight on my part.  `BlobUriBuilder.BlobName` and `.Path` should now be escaped, based on another PR I merged.  I'll address now.

**Pull request merged**